Respect the credentials mode (in section 2.3. Fetch integration)

msramek opened the pull request:
This resolves #4.

mikewest commented on index.src.html:
```diff
@@ -519,14 +520,18 @@ <h3 id="fetch-integration">Fetch Integration</h3>
   execute the following step:
```
Nit: Can you change the line above this to refer to "step #12" while you're here?

msramek replied:
Done.

mikewest commented on index.src.html:
```diff
@@ -519,14 +520,18 @@ <h3 id="fetch-integration">Fetch Integration</h3>
-      <a>`Clear-Site-Data`</a>, then execute [[#clear-response]] on
-      |response|.
+      <a>`Clear-Site-Data`</a> and the request's <a>credentials mode</a> permits
+      modifying credentials, then execute [[#clear-response]] on |response|.
```
We should be more specific here: perhaps "`<a>credentials mode<a>` is `include`, or `<a>credentials mode</a>` is `same-origin` and its <a>response tainting</a> is `basic`"?

mikewest replied:
(Note that this would be simpler if we moved the hook to [HTTP-Network fetch](https://fetch.spec.whatwg.org/#http-network-fetch), which would simplify the service worker bits as well.)

annevk replied:
I don't really see why the entire header is conditional upon this. By doing that we extend "credentials mode" to also affect things it didn't previously affect, such as `localStorage` and co. I guess it makes some sense to do that, but I'm a little wary about it.

msramek replied:
My reasoning is that the intent of restricting credentials during `fetch()` is that "I only want to load some content, I don't want modifications to my data" so applying this to other data types seems natural. It's true that the definition of `credentials mode` only pertains to cookies, but then again, it was written at a time when there were no headers affecting `localStorage` (...at least to my knowledge?).

But I don't feel strongly either - happy to defer to you.

Regarding the other point, to make sure I understand - should I close this pull request and make a change in the fetch spec? It just didn't seem right to me to modify an established spec with an experimental feature, but if it's OK, I'll do that.

msramek replied:
@annevk SGTM. So I'll leave it here for now.

@mikewest Updated. Changed it to bulletpoints as it was difficult to parenthesize the boolean operators in that sentence. When we move this to the `fetch` spec, it seems that we can shorten it to "has a credentials flag set".

mikewest commented:
@annevk probably has thoughts here as well.

mikewest commented:
If you move this algorithm (as suggested in https://github.com/w3c/webappsec-clear-site-data/pull/10), then I think this PR boils down to looking at the `credentials flag`, right?

msramek commented:
@mikewest Yes, I mentioned that in the comment thread above :)

Well, it's up to you. If you want me to move it to the fetch spec (and @annevk agrees), I can do so, even though I'm not sure we're ready for that with just one implementation. Note that that also means removing the entire section 2.3.

And in that case, to clarify, we probably want network-only (as opposed to network-or-cache), right?

mikewest commented:
> If you want me to move it to the fetch spec (and @annevk agrees)

When I typed "move", I meant that I want you to monkey-patch a different algorithm in Fetch than we're doing today. I agree with @annevk that we should hold off on patching Fetch until there's more support from vendors other than Google.

In other words, I'd suggest that you change section 2.3 to something like:

```
If the `Clear-Site-Data` header is present in an HTTP response received from the network, then data MUST be cleared before rendering the response to the user. That is, after step 14 of the current "HTTP-network fetch" algorithm, execute the following step:

15.  If |credentials flag| is set, and |response|'s header list contains a header named Clear-Site-Data, then execute "Clear data for response" on |response|.
```

msramek commented:
...I was OOO in the meantime and couldn't get back to this until now, sorry.

Agreed that moving this to network fetch is a nice way to also close #10. I updated the text and also added the pedantic reference to [[CLEAR-SITE-DATA]] which will have to be there once we copy over the text to the fetch spec :-)

mikewest commented on index.src.html:
```diff
@@ -513,20 +514,24 @@ <h3 id="fetch-integration">Fetch Integration</h3>
-      |response|.
+  15.  If |response|'s `credentials flag` is set, and |response|'s
+       <a>header list</a> contains a header named <a>`Clear-Site-Data`</a>, then
+       execute [[#clear-response]] of [[CLEAR-SITE-DATA]] on |response|.
```
1.  `credentials flag` isn't a property of `response`: it's an argument to the "HTTP-network fetch" algorithm. How about:

  ```
If <var ignore>credentials flag</var> is set, and |response|'s
<a>header list</a> contains a header named
<a>`Clear-Site-Data`</a>, then execute [[#clear-response]]
on |response|.
  ```

  Note the `<var ignore>` bit, which just landed in Bikeshed, and lets us mark something up as a monkey-patched variable without it giving us annoying warning messages.

2. Please drop the `[[CLEAR-SITE-DATA]]` bit; I agree that we'll want to make that reference when we upstream this to Fetch, but for the moment, it's confusing Bikeshed's tiny brain and adding a reference to this document _to_ this document, which is odd. :)

msramek replied:
Done. Sorry, didn't notice that :)

I updated bikeshed, but still see the same warning regarding the <var>. I guess the change hasn't been rolled out yet? Well, if it will soon, we can still proceed, since it's a non-fatal warning.

mikewest commented on index.src.html:
```diff
@@ -513,20 +514,24 @@ <h3 id="fetch-integration">Fetch Integration</h3>
-  the user. That is, before step #9 in the current <a>main fetch</a> algorithm,
-  execute the following step:
+  If the <a>`Clear-Site-Data`</a> header is present in an HTTP <a>response</a>
+  received from the network, then data MUST be cleared before rendering the
```
Would you mind adding a note somewhere (perhaps in a "Service Workers" section under "Security Considerations") that outlines the rationale for only accepting responses from the network, and not those from Service Workers (e.g. cached responses, synthesized responses, etc).

msramek replied:
Done.

I copied over and slightly adapted the text from the other pull request which I closed. Let me know what you think.

By the way, should we also elaborate on cache? The fact that this is network-only means that the standard HTTP cache is also ignored (and I don't mean just service workers' caches.match() ).

mikewest replied:
The new text is reasonable. It's probably reasonable to mention cache as well. If you feel so inclined, I'd be happy to look at more text. :)

mikewest commented:
I'll give @annevk a bit to weigh in if he's interested, and land this tomorrow. Thank you for going through a few iterations!

msramek commented:
Thanks!

I responded to the other comment asynchronously (and later than your LGTM) and have another question there - please have a look :) But we can also address that in another pull request.
